clang/test/CIR/CodeGen/array-ctor.cpp

```diff
@@ -0,0 +1,70 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s -check-prefix=CIR
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -fclangir -emit-llvm %s -o %t-cir.ll
+// RUN: FileCheck --input-file=%t-cir.ll %s -check-prefix=LLVM
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -emit-llvm %s -o %t.ll
+// RUN: FileCheck --input-file=%t.ll %s -check-prefix=OGCG
+
+struct S {
+    S();
+};
+
+void foo() {
+    S s[42];
+}
+
+// CIR: cir.func dso_local @_Z3foov()
+// CIR:   %[[ARRAY:.*]] = cir.alloca !cir.array<!rec_S x 42>, !cir.ptr<!cir.array<!rec_S x 42>>, ["s", init]
+// CIR:   %[[CONST42:.*]] = cir.const #cir.int<42> : !u64i
+// CIR:   %[[DECAY:.*]] = cir.cast(array_to_ptrdecay, %[[ARRAY]] : !cir.ptr<!cir.array<!rec_S x 42>>), !cir.ptr<!rec_S>
+// CIR:   %[[END_PTR:.*]] = cir.ptr_stride(%[[DECAY]] : !cir.ptr<!rec_S>, %[[CONST42]] : !u64i), !cir.ptr<!rec_S>
+// CIR:   %[[ITER:.*]] = cir.alloca !cir.ptr<!rec_S>, !cir.ptr<!cir.ptr<!rec_S>>, ["__array_idx"]
```

Comment:
This alloca should be with the other allocas at the start of the function.

Reply:
If anything the alloca should be at the start of the region its in as we might create the alloc in a scope nest but is there a reason beyond readability?

Reply:
The alloca location determines when stack space is reserved and isn't generally concurrent with the lifetime of the variable. As a general rule, all allocas for the function are gathered at the top of the entry block, regardless of the scope of the variables they represent. Some targets aren't tolerant of allocas in the middle of a function. Classic codegen maintains a variable (AllocaInsertPt) in CodeGenFunction for this exact purpose. We don't seem to have one of those in CIR codegen yet.

Reply:
I just realized my thinking about this was a bit off. In CIR, we do place the alloca op in the region where the variable is logically allocated. It is hoisted by a later pass to where LLVM IR will expect it. So, you were right that it shouldn't be at the start of the function for nested scopes.